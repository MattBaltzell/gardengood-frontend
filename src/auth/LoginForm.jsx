import React from "react";
import useFormFields from "../hooks/useFormFields";
import { useNavigate, Link } from "react-router-dom";
import "../components/Form/Form.css";

const LoginForm = ({ login }) => {
  const INITIAL_STATE = { username: "", password: "" };
  const { formData, setFormData, handleChange } = useFormFields(INITIAL_STATE);

  const navigate = useNavigate();

  const handleSubmit = (e) => {
    e.preventDefault();
    login(formData);
    setFormData(INITIAL_STATE);

    navigate("/");
  };

  return (
    <main className="Form">
      <div className="Home__bg"></div>

      <h1>Login</h1>
      <form onSubmit={handleSubmit}>
        <label htmlFor="username">Username</label>
        <input
          id="username"
          name="username"
          type="text"
          onChange={handleChange}
          value={formData.username}
          autoComplete="username"
        />
        <label htmlFor="password">Password</label>
        <input
          id="password"
          name="password"
          type="password"
          onChange={handleChange}
          value={formData.password}
          autoComplete="current-password"
        />
        <button>Submit</button>
      </form>
      <p>
        No Account? <Link to="/signup">Create one</Link>
      </p>
    </main>
  );
};

export default LoginForm;
