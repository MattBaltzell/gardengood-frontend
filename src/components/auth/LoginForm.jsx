import React, { useContext, useState } from "react";
import { Link, Navigate } from "react-router-dom";
import { Formik, Form } from "formik";
import * as Yup from "yup";
import { MyTextInput, MyPasswordInput } from "../form/Forms";
import "../form/Form.css";
import UserContext from "./UserContext";
import { v4 as uuid } from "uuid";

const LoginForm = ({ login, toast }) => {
  // Pass the useFormik() hook initial form values, a validate function that will be called when
  // form values change or fields are blurred, and a submit function that will
  // be called when the form is submitted

  const currUser = useContext(UserContext);
  const [submissionErrors, setSubmissionErrors] = useState([]);

  const INITIAL_STATE = {
    username: "",
    password: "",
  };

  return currUser ? (
    <Navigate to="/" />
  ) : (
    <main className="Form">
      <div className="Home__bg"></div>

      <h1>Login</h1>

      <Formik
        initialValues={{ ...INITIAL_STATE }}
        validationSchema={Yup.object({
          username: Yup.string()
            .max(25, "Must be 25 characters or less")
            .required("Required"),
          password: Yup.string()
            .min(8, "Must be 8 characters or more")
            .required("Required"),
        })}
        onSubmit={(values) => {
          setSubmissionErrors([]);
          setTimeout(async () => {
            const res = await login(values);
            if (res.error) {
              setSubmissionErrors((errors) => [...errors, res.error.message]);
            }
          }, 400);
        }}
      >
        <Form>
          <MyTextInput
            label="Username"
            name="username"
            type="text"
            autoComplete="username"
          />
          <MyPasswordInput label="Password" name="password" type="password" />
          {submissionErrors
            ? submissionErrors.map((e) => (
                <div key={uuid()} className="error center">
                  {e}
                </div>
              ))
            : null}
          <button type="submit">Log In</button>
        </Form>
      </Formik>

      <p>
        No Account? <Link to="/signup">Create one</Link>
      </p>
    </main>
  );
};

export default LoginForm;
