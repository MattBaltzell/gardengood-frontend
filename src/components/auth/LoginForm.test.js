import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import LoginForm from "./LoginForm";

describe("LoginForm component tests", function () {
  it("renders without crashing", function () {
    render(
      <MemoryRouter>
        <LoginForm></LoginForm>
      </MemoryRouter>
    );
    expect(screen.getByText("Username")).toBeInTheDocument();
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <LoginForm />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("submits");
});
