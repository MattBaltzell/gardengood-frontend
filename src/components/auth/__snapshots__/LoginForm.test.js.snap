// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component tests matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="Form"
  >
    <div
      class="Home__bg"
    />
    <h1>
      Login
    </h1>
    <form
      action="#"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        autocomplete="username"
        class="text-input"
        name="username"
        type="text"
        value=""
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="password-input"
        name="password"
        type="password"
        value=""
      />
      <button
        type="submit"
      >
        Log In
      </button>
    </form>
    <p>
      No Account? 
      <a
        href="/signup"
      >
        Create one
      </a>
    </p>
  </main>
</DocumentFragment>
`;
